### PR TITLE
[FluidDynamicsApplication] [MPI] Trilinos MPI version of TwoFluidSolver

### PR DESCRIPTION
**Pull request by swenczowski**
The following implementations are suggested:

- Introduction of an MPI variant (based on Trilinos) of the TwoFluids python solver.
- An MPI version of the the process used to check the mass conservation inside the domain.

### CODE REVIEW

**Comment by rubenzorrilla** [comment]
Before any deeper review, is it possible to have a unique process to check the mass conservation?

**Comment by swenczowski** [comment]
Judging from the definition of "apply-outlet-process", it should be possible. I will try my best...

File: applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py
Comment (jcotela) [inline applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py]: You don't seem to need `math` in this file (or I can see no uses of the module)

File: applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py
Comment (jcotela) [inline applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py]: I would recommend using the `_is_printing_rank` check for uniformity.

File: applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py
Comment (jcotela) [inline applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py]: note that, when using a `with` statement, you don't need to close the file explicitly. It will be closed automatically when exiting the context of the `with`, even if something throws an error (this is the whole advantage of the `with`).

**Comment by swenczowski** [comment]
I am sorry for the confusion and the "forced push". By mistake, I deleted the wrong file directly from the repository. Anyway, now there is one "apply_mass_conservation_check_process" for serial and MPI applications.

File: applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py:47]: to be honest i don't love this name for the solver

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py]: is this barrier really needed?

File: applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py
Comment (jcotela) [inline applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py:47]: I would go with `TwoFluid`. It is easier to maintain if the name somehow matches the python class/files

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py
Comment (jcotela) [inline applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py]: you have a `consider_periodic_conditions` flag above. How is it different from this parameter?

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py
Comment (jcotela) [inline applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py]: Since this is a new solver, I would try to follow the style of the recent fluid and embedded solvers for the parameter names and default values. Please check the `vmsmonolithic` or `embedded` versions (in particular here they have a specific block for stabilization-related settings such as the `dynamic_tau`.)

**Review by jcotela — COMMENTED** [review]
Overall good work. I have some relatively minor comments.

File: applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py]: Done. Thank you

File: applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py]: I agree. Done.

File: applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py]: I removed it - hopefully I will keep it in mind now. Thank you.

File: applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py:47]: Yes, I agree in this aspect - I often forget or add the "s" by mistake. The reason why I chose this keyword was because it coincides with the one used for the OpenMP version of the solver. I assume it would make sense to use the same keyword in both cases, wouldn't it? 
The serial solver class is called `class NavierStokesTwoFluidsSolver(FluidSolver)` by the way, where the "s" is present. Also the serial python file itself is named `navier_stokes_two_fluidS_solver.py`. Maybe it is not so bad to maintain the "s" and to alter the name of the Trilinos file and class name, instead.
@jcotela Would that be an option for you?
@rubenzorrilla May I also ask for your opinion?

Comment (jcotela) [reply applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py:47]: I would be OK with that, if it means that we have a consistent naming scheme (C, python, json, serial and trilinos).

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py]: I am very sorry, but I am not entirely sure if the corrections are going into the right directions:
- "dynamic_tau" moved to "formulation"
- "AmgclMPISolver" set as standard solver (as recommended in a meeting)
- apart from that, the outline is quite similar to the "embedded" version (actually served as the reference)
It would be very kind if you could have another look. Sorry in case I missed further important aspects.

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py]: After investigating closer, it seems that this entry is no longer longer needed or even provokes "depreciated" messages in some cases. Thus, it is removed. In contrast, the entry "consider_periodic_conditions" remains since it is vital for the domain decomposition.

File: applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py:47]: As @jcotela said, any option is valid provided that it is consistent with the rest of files. So I'm OK as well.

File: applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py]: I would put the file name as a parameter of the default JSON settings below so a user-defined file name can be specified if needed (e.g. imagine that you want to compare the mass conservation of multiple runs).

**Comment by rubenzorrilla** [comment]
@swenczowski please solve the conflicts with applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py. This has been already fixed by @philbucher so it should be as simple as keeping the changes of the master (incoming).

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py
Comment (philbucher) [inline applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]: Yep please revert the changes in this file

File: applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py:47]: Alright. Now everything is named "...TwoFluids..." with "s"

File: applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/apply_mass_conservation_check_process.py]: Changed to a parameter in the *.json file.

File: applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py
Comment (swenczowski) [reply applications/trilinos_application/python_scripts/trilinos_import_model_part_utility.py]: Yes, of course - just tried to get around the problem quickly. I hope this problem is now solved after I merged the master and accepted all incoming changes for this file.

File: scripts/build/nightly/configure_clang.sh
Comment (philbucher) [inline scripts/build/nightly/configure_clang.sh]: You should not delete those files!

Comment (swenczowski) [reply scripts/build/nightly/configure_clang.sh]: Ohh, I thought them to be local files that somehow escaped my git-ignore settings. I am very sorry. I will try and revert this immediately.

**Comment by swenczowski** [comment]
@philbucher Hopefully the problems I caused are solved with these operations. Again, my apologies for the alarm and the mess. I was really convinced that these files should not be there...

File: applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp:62]: just a hint: you could use here the other constructor. (what you did is ok, just take this as a hint for the future)

File: applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py
Comment (RiccardoRossi) [reply applications/FluidDynamicsApplication/python_scripts/python_solvers_wrapper_fluid.py:47]: ok

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py]: I also was not sure about this one, but in the end I think it makes sense:
AddDofs() is the last function call in the constructor of the class AnalysisStage. Thus, the barrier assures, that all processes have finished the construction and all variables are available. E.g. in my simulation cases, the next function call would be "simulation.Run()". If any process has not yet established all variables, this could cause unpredictable behaviour.

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py
Comment (philbucher) [inline applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:100]: I dont think those barriers are needed
Afaik the other solvers dont have it
Simon did you have problems?

Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:100]: For the four processes on my laptop it seems to work without the barrier.
However: A comparable structure concerning the arrangement of barriers is used in TrilinosNavierStokesSolverMonolithic, NavierStokesMPIEmbeddedMonolithicSolver, TrilinosNavierStokesSolverFractionalStep, and AdjointVMSMonolithicMPISolver. Also from a look at the code, I think that leaving them out could cause a lot of trouble. Thus, I believe that one should not rely on a experimental result on a 4-process machine and that the barriers should remain. Though the "but-it-does-not-cost-so-much" argument is not the best, one could consider that this function is not repeated in an innermost loop but only called once for a simulation run.

Comment (philbucher) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:100]: Hm ok

Fine for me, maybe @jcotela knows more here
I‘d be interested to hear his opinion

**Comment by swenczowski** [comment]
After the discussion seems to have reached a certain standstill, I would like to kindly ask if there are any further hints for improvement of the suggested implementation, @jcotela @rubenzorrilla? 
The reason for the question is that I am getting confused with all the currently open (local) branches and merging this PR would help me to organize my further development. Thank you very much for your answer in advance.

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py
Comment (jcotela) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:100]: It is true that most MPI fluid solvers have it (probably my fault) but in fact I am not so sure it is needed.

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py
Comment (jcotela) [inline applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:76]: Note that you are not doing what the comment says you are doing.

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py
Comment (jcotela) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluid_solver.py]: the new settings look OK to me.

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:76]: Thank you very much and sorry for missing this obvious one. I remember that I did not feel confident altering this and, thus, probably left the comment.

Comment (jcotela) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:76]: I think it may be worth to consider why you would want to do that. I did not check in detail, but I can tell you that the monolithic trilinos solver skips one level of constructor to avoid doing the validation of input parameters twice. I also make sure not to initialize member variables twice (self.something = ... assigned both by your class and the base class, probably to different values). If you can be sure you are not doing this kind of things, you should be fine.

Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:76]: Exactly the double assignment of member variables was my concern as well - and yes, it is unfortunately the case. At the same time, the Codacy check validated `super( parentClass, self )`  as an error. Finally, I did internet research and found that the attribute values set by the parent class constructor should be over-written. I entirely understand that this is not nice or good practice but for me it solved the complaints coming from Codacy. Sorry for this pragmatically stupid thinking. Is there a way to restore the old `super( parentClass, self )`-version and still avoid the problems with Codacy?

Comment (jcotela) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:76]: If your main concern is to silence codacy, you could probably fool it by having your base class constructor call a class method (move everything from the constructor to that method) and reimplement that method in the derived class (I never tried it, just a suggestion). Personally, I do not think it is worth the cost in readability, but your call. However, keep in mind that codacy is a bit over-sensitive, and there are some instances where the things it suggests are simply not practical for Kratos (I believe it focuses too much in pure python, while we have some constraints imposed by the C interface, such as having to import some applications from files that do not specifically use them).

Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:76]: I agree with you in this aspect that a change (moving the content of the constructor to another function) would deteriorate the readability. Is there a way for me to merge the pull request without "silencing" Codacy by fulfilling its remarks? If there is one, I would clearly prefer it.

Comment (jcotela) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:76]: Not that I know of, but you don't need to "pass" codacy to merge (we use it as a guideline, but not passing it will not block the PR, unlike travis or appveyor).

Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:76]: Thank for this hint - I was not aware of the fact that it does not have the "required" label. Thus, the change of [the questionable commit](https://github.com/KratosMultiphysics/Kratos/pull/3524/commits/d92002abbf10fb6bff24293769d35956bd6587b4) was reverted and the comment was restored.

File: applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py
Comment (swenczowski) [reply applications/FluidDynamicsApplication/python_scripts/trilinos_navier_stokes_two_fluids_solver.py:100]: Would you agree if we keep the barrier for now, @jcotela @philbucher ? At the same time, I promise that I will keep the question in mind. After merging, the implementation could by tested on the cluster which could give as more reliable results.

**Review by rubenzorrilla — APPROVED** [review]

**Pull request merged**
